pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,542 @@
+			}
+		}
+		return offsets, nil
+	}
+
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 2 * time.Second,
+		MaxRetries: 10,
+	})
+	for boff.Ongoing() {
+		var offsets map[int32]kgo.Offset
+		offsets, err = fetchOffsets(ctx)
+		if err == nil {
+			return offsets, fallbackOffset, nil
+		}
+		level.Warn(b.logger).Log("msg", "failed to fetch startup offsets", "err", err)
+		boff.Wait()
+	}
+	// Handle the case the context was canceled before the first attempt.
+	if err == nil {
+		err = boff.Err()
+	}
+	return nil, -1, err
+}
+
+func commonKafkaClientOptions(cfg KafkaConfig, logger log.Logger, metrics *kprom.Metrics) []kgo.Opt {
+	opts := []kgo.Opt{
+		kgo.ClientID(cfg.ClientID),
+		kgo.SeedBrokers(cfg.Address),
+		kgo.DialTimeout(cfg.DialTimeout),
+		kgo.MetadataMinAge(10 * time.Second),
+		kgo.MetadataMaxAge(10 * time.Second),
+		kgo.WithLogger(newKafkaLogger(logger)),
+	}
+	if metrics != nil {
+		opts = append(opts, kgo.WithHooks(metrics))
+	}
+	return opts
+}
+
+func (b *BlockBuilder) waitKafka(ctx context.Context) error {
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Second,
+		MaxRetries: 0,
+	})
+	for boff.Ongoing() {
+		err := b.kafkaClient.Ping(ctx)
+		if err == nil {
+			return nil
+		}
+
+		level.Error(b.logger).Log(
+			"msg", "waiting for kafka ping to succeed",
+			"part", fmt.Sprintf("%v", b.parts),
+			"num_retries", boff.NumRetries(),
+		)
+		boff.Wait()
+	}
+	return boff.Err()
+}
+
+func (b *BlockBuilder) stopping(_ error) error {
+	b.kafkaClient.Close()
+
+	return nil
+}
+
+func (b *BlockBuilder) running(ctx context.Context) error {
+	// Do initial consumption on start using current time as the point up to which we are consuming.
+	// To avoid small blocks at startup, we consume until the last hour boundary + buffer.
+	cycleEnd := cycleEndAtStartup(b.cfg.ConsumeInterval, b.cfg.ConsumeIntervalBuffer)
+	err := b.NextConsumeCycle(ctx, cycleEnd)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return err
+	}
+
+	nextCycleTime := time.Now().Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+	waitTime := time.Until(nextCycleTime)
+	for waitTime > b.cfg.ConsumeInterval {
+		// NB: at now=14:12, next cycle starts at 14:15 (startup cycle ended at 13:15)
+		//     at now=14:17, next cycle starts at 15:15 (startup cycle ended at 14:15)
+		nextCycleTime = nextCycleTime.Add(-b.cfg.ConsumeInterval)
+		waitTime -= b.cfg.ConsumeInterval
+	}
+
+	for {
+		select {
+		case <-time.After(waitTime):
+			cycleEnd := nextCycleTime
+			level.Info(b.logger).Log("msg", "triggering next consume from running", "cycle_end", cycleEnd, "cycle_time", nextCycleTime)
+			err := b.NextConsumeCycle(ctx, cycleEnd)
+			if err != nil {
+				b.metrics.consumeCycleFailures.Inc()
+				level.Error(b.logger).Log("msg", "consume cycle failed", "cycle_end", cycleEnd, "err", err)
```

Comment:
TBF, I'm not sure what is the desired behaviour when an error reaches the `NextConsumeCycle`. Should we crash the whole app? 

Note that the pieces of the business logic, that 
1. consume from Kafka,
2. build blocks,
3. upload blocks,
4. update consumer's offset and commit the offset

are parts of the future PRs.

Reply:
what we do with regular consumption in the ingester is retry indefinitely. It's also dangerous because the app can block forever, but it's harder to lose data. There'd be some errors which you can ignore and others on which you always retry. I think you already have the error classification in another PR that I've reviewed

Reply:
I think the best we can do here is to fail the application, similar to how ingest storage does it ([code](https://github.com/grafana/mimir/blob/fdb1225907b237a4ffd4cbd6950385e87d71f2cf/pkg/storage/ingest/reader.go#L173-L181)). Everything that can handled gracefully (by retrying or logging) is handled inside `NextConsumeCycle`